universe/NamedValueRefManager.cpp

```diff
@@ -198,35 +201,78 @@ template void RegisterValueRef(std::string name, std::unique_ptr<ValueRef::Value
 ///////////////////////////////////////////////////////////
 namespace ValueRef {
 template <typename T>
-NamedRef<T>::NamedRef(std::string value_ref_name) :
-    m_value_ref_name(value_ref_name)
+NamedRef<T>::NamedRef(std::string value_ref_name, bool is_lookup_only) :
+    m_value_ref_name(value_ref_name),
+    m_is_lookup_only(is_lookup_only)
 {
-    DebugLogger() << "ctor(NamedRef<T>): " << typeid(*this).name() << " value_ref_name: " << m_value_ref_name;
+    TraceLogger() << "ctor(NamedRef<T>): " << typeid(*this).name() << "  value_ref_name: " << m_value_ref_name << "  is_lookup_only: " << m_is_lookup_only;
+}
+
+
+// waits for named ref to be registered and initializes invariants
+// called on first use, necessary for derived values/caching
+template <typename T>
+bool NamedRef<T>::NamedRefInitInvariants()
+{
+    if (m_invariants_initialized)
+        return true;
+    auto* vref = GetValueRef();
+    if (!vref) {
+        std::chrono::milliseconds msecs(200);
+        InfoLogger() << "NamedRef<T>::NamedRefInitInvariants() could not find value ref, will sleep a bit and retry.";
+        std::this_thread::sleep_for(msecs);
+        vref = GetValueRef();
+        for (int = 2; i < 5; i++) {
```

Comment:
`i` was missed. Also multiple tries looks fishy. Can we have it more determined?

Reply:
> `i` was missed. Also multiple tries looks fishy. Can we have it more determined?

if i can figure out a way to do the intialisation after parsing we are basically done
my problem is that the AppInterface/ServerApp::StartBackgroundParsing (which knows when parsing the contents is finished) does not have references to the created value refs. Probably could add a function like InitialiseAfterParsingContents to those managers which would relay that to their content type objects

the question i want now to answer, if your mismatches stop if no "Giving up" happens.

Reply:
so the answer is, yes mismatches still occur even if all and i found that the value refs from xenophobic.macros did not register on the AI client (WTF?). For getting feedback I removed the NamedReal value refs from the macro file.

Anyway this means the remaining mismatches are unrelated to this PR.